### ISSUE TITLE
Bazel rules

### ISSUE THREAD

**Issue by amckinney** [bufbuild/buf#681]
This serves as the parent task for all of the work related to publishing official Bazel rules.

We've had plans for official Bazel rules for a while now ([roadmap](https://docs.buf.build/roadmap#bazel-rules)), but we're not sure exactly what they'll look like in practice. Will they be specific to lint and breaking like [rules_proto_grpc](https://rules-proto-grpc.com/en/latest/lang/buf.html), or will they cover other use cases like `buf generate`?

@shawntabai did some exploration around what it would look like to either reuse `[go_]proto_library` or re-create our own Buf-specific rules in https://github.com/shawntabai/bufzel.

**Note that the primary goal was to see whether or not they could leverage Managed Mode within the context of Bazel.**

Their findings are summarized by the following:

* The main benefit of any of this is managed mode, since the linter and breaking change detector are already available separately.
* Buf in Bazel should isolate protos into small buildable units (the way Bazel normally does).
* The standard proto_library probably doesn’t benefit much from being made to work with Buf and can probably be used without modification.
* The standard language-specific generator rules (e.g. go_proto_library) don’t seem flexible enough to make reusing them with Buf feasible
    * Buf generation rules could potentially be a wrapper around these that specify custom compilers based on their options, but that seems more complicated than it’s worth.
    * Buf generation rules should be able to change plugins, but not the core compiler. so they should be forced into using the Buf compiler, while all additional functionality should be added onto that in the form of protoc plugins (the default rules appear to lump all compilers together in a way that makes not using protoc very awkward).
* It’s questionable if the work is truly worth it, if the main goal is to support Managed Mode. There could very well be an easier way to do that.

We'll use this task to continue discussing what needs to be done for Buf's Bazel rules and what use cases it should cover in general.

**Comment by amckinney** [comment]
@robbertvanginkel and I discussed this a bit more and came away with the following things to look into (in order of priority):

## Lint

We should provide a mechanism for running `buf lint` (a.k.a `protoc-gen-buf-lint`), but it can take a variety of different forms.

In [rules_proto_grpc](https://rules-proto-grpc.com/en/latest/lang/buf.html), you'll notice that they write `buf lint` rules as test rules (i.e. `buf_proto_lint_test`). This is because test rules continue to execute in the midst of failures, whereas standard rules will stop execution as soon as a build rule fails.

The test rule approach is acceptable, but it's worth exploring the [Bazel aspect](https://docs.bazel.build/versions/main/skylark/aspects.html) approach (which is preferred if we can get it right).

  * This makes it so that we could take the output from a compile step (i.e. a `FileDescriptorSet` via [proto_library](https://docs.bazel.build/versions/main/be/protocol-buffer.html#proto_library)), then run `protoc-gen-buf-lint` on that output.
  * Aspects also make it so that you don't need to change your build rules - you only need to change the `bazel build ...` invocation. There's an example of a `bazel build ...` invocation with aspects [here](https://github.com/thundergolfer/bazel-linting-system) (note that we should not actually use `bazel_linting_system` - it's only meant to be an example for aspects).

## Breaking

`buf breaking` is a little more challenging than `buf lint` since it requires an input to compare `--against`. For Bazel, it's difficult to obtain a compiled representation for a `FileDescriptorSet` / `Image` from external sources like `.git`, which is our traditional use case (re: [tour](https://docs.buf.build/tour/detect-breaking-changes)).

One approach that was quickly rejected was to build a way to compile a `FileDescriptorSet` from an external Bazel workspace. In this case, the external workspace could be a reference to the user's GitHub repository so that they can compare against the previous commit. However, this approach could be _very_ expensive, and would potentially violate some of Bazel's remote caching requirements, as well as be prone to deeply nested / recursive Bazel workspace builds.

Alternatively, our rule could require that users plug in a `FileDescriptorSet` / `Image` manually like [buf_proto_breaking_test](https://github.com/rules-proto-grpc/rules_proto_grpc/tree/master/buf/example) does.

Finally, we can explore if users should compare compatibility `--against` module references. For example, the `buf breaking` invocation can be imagined like so (where the latest commit is being compared against):

```sh
$ buf breaking <input> --against buf.build/acme/petapis:84a33a06f0954823a6f2a089fb1bb82e
```

This would require that the rule is able to use `buf` to download the module, which is similar to the next point of exploration.

## Module Dependencies

We currently don't have a way to retrieve module dependencies (i.e. `deps:` specified in the user's `buf.yaml`). We might be able to take inspiration from how the rules for Go modules are implemented. This might require a [Gazelle](https://github.com/bazelbuild/bazel-gazelle) extension to know how to resolve external imports into `proto_library` rule `deps`.

## Managed Mode

As discussed in some of the comments above, the `buf generate` experience is somewhat at odds with Bazel's architecture. Bazel is inherently more granular, requiring invocations on individual files, not buildable units as a whole. There might be something we can do to provide a `buf generate` experience, but the differentiating factor in a Bazel environment (which already has the entire build graph laid out), is Managed Mode.

This ties back into the comments and prototype that Shawn put together above. But it's worth exploring whether or not we could build a `FileDescriptorSet` modifier rule that simply applies Managed Mode to a pre-compiled `FileDescriptorSet`. With this, users would be able to replace their instance of `proto_library` with one that both compiles with `protoc`, then modifies the result according to their Managed Mode configuration.

We could either replace their implementation of `proto_library` with a [Bazel Macro](https://docs.bazel.build/versions/main/skylark/macros.html) so that it expands to multiple actions (instead of just compiling with `protoc`), or have users replace their instances of `proto_library` with something like a `buf_proto_library`.

Regardless, the experience we want to look into is a rule that takes a `FileDescriptorSet` as input, and returns a `FileDescriptorSet` as output. Depending on whether or not this is actually what we end up doing, we'll need a `buf` command that can do this.

--

TL;DR - We're convinced that we'll need a solution for `buf lint` and `buf breaking` in Bazel environments. The `buf build` and `buf generate` commands are a little less clear, but we know that it'd be valuable for users to leverage Managed Mode, so we should see how we can do so without needing to build out `buf generate` and/or `buf protoc` functionality for Bazel because it's otherwise susceptible to significant complexity (re: Gazelle automation caveats in the comments mentioned above).

**Comment by srikrsna-buf** [comment]
Are we planning to expose download links on BSR for image files against each commit?

If we can do that, breaking change detection can just use `http_file` to get the first version of current major version (locked by using the commit hash) and check against that. The commit hash need not change until the next major version bump.

**Comment by amckinney** [comment]
> Are we planning to expose download links on BSR for image files against each commit?

This seems like a valuable feature in general, but we don't expose that functionality over HTTP/1 yet. Fortunately, users can already get the image associated with a specific commit with a command like the following:

```
$ buf build buf.build/acme/petapis:<$commit> -o image.bin
```

Although `http_file` seems like a simple drop-in for Bazel, I wonder if there's something else we can do to invoke `buf` like the above in the interim (i.e. while we don't expose this functionality over HTTP/1).

> If we can do that, breaking change detection can just use `http_file` to get the first version of current major version (locked by using the commit hash) and check against that. The commit hash need not change until the next major version bump.

Buf modules aren't always meant to conform to the same semantic versioning scheme we see for other software dependencies in general (e.g. `v1.0.0`). In short, we aren't sure that we even want to encourage users to tag their BSR repositories with semantic versions since it shouldn't be necessary with respect to consuming APIs. With the BSR, we instead want to encourage users to enforce compatibility within a release track (formerly known as branches, where `main` is the default).

So ideally we would be able to simply just check against the HEAD commit on `main` (which is the default behavior when a `<$commit>` is omitted from the command above), or their release track of choice (which could be a configuration option for the rule).

With all that said, I wonder if there's a way for us to emulate the `http_file` behavior that you're suggesting, but define a rule that uses the `buf build` command I've shown above instead.

**Comment by robbertvanginkel** [comment]
> If we can do that, breaking change detection can just use http_file to get the first version of current major version (locked by using the commit hash) and check against that

`http_*` rules can only be locked with hashes over the content of what the http endpoint returns, like a sha256 hash: https://docs.bazel.build/versions/main/repo/http.html#http_archive-sha256. As bazel's goal is generally byte by byte reproducible builds having a hash that isn't guaranteed to return exactly the same content is frowned upon.

The `go_repository` rules do show that it's possible to implement a different rule that uses a different type of checksum. There is some downsides to this though:
- This circumvents the remote asset api - which takes care of downloading/caching/proxying external dependencies. I believe this one requires a standard hash over the downloaded content.
- I think the go sum contract is that once a module/version has been created and summed, it will never return different content. I don't think we have the same guarantees for images, perhaps we do have it for downloading only the files (but I don't think that solves the problem of needing an input for buf breaking.

**Comment by Dogild** [comment]
Thanks for tackling this problem! I am looking forward for this feature - please let me know if I can help somehow.

**Comment by stabai** [comment]
Just wanted to chime in on code generation, if that's still being considered / worked on. As I've worked with Go more, I've discovered how incompatible the Go compiler is with Bazel's model of compiling protos to its own output folders. The Go compiler expects the generated code to either be colocated with the rest of your code, or in a separate formal module. Since `protoc-gen-go` doesn't seem to have an option to also generate a Go module, and Bazel strictly prohibits any `build` actions from altering the workspace, this means that things can't work properly. This would be OK as long as you always use Bazel and don't need to invoke the Go compiler directly, but this also breaks `golangci-lint` (because of its dependence on the Go compiler), which is far more problematic. I'm under the assumption this is only an issue with Go because I haven't encountered this sort of thing with any languages other than Go, but my experience with protos is pretty limited outside of Java/Kotlin and JavaScript/TypeScript.

So I have a couple of possible suggestions for how to fix this for Go. The first is to have the proto code generator for Go also create a Go module. Since Buf typically processes entire directories together, this may actually be very simple, but I'm not sure. All this would require from the user is adding a `replace` directive in their main module to point to the output directory generated by Buf's Bazel rule. Problem solved!

Another possible solution (which is the best solution I've seen with the current `go_proto_library` rule), is to also have a [clever sort of macro](https://github.com/stackb/rules_proto/blob/7c95feba87ae269d09690fcebb18c77d8b8bcf6a/rules/proto_compiled_sources.bzl) which duplicates the output into the workspace's source tree. It has to do this by actually creating 3 targets:
1. a `build` target that generates the code into Bazel's output directory
2. a `run` target that copies the generated code from Bazel's output directory to the workspace source tree (run targets are the only ones that can do this)
3. a `test` target that ensures the two match (and if they don't, suggests running step 2 to fix it)

This makes a pretty decent solution since any sensible CI setup with Bazel will involve running those tests. Hope this helps!

**Comment by stabai** [comment]
Is there any update on the roadmap for this work? I'm at a point where I'm going to have to solve this soon, so I'm trying to decide if I should use a workaround until the official solution is ready, or take the time to make some good rules that others can also use.

**Comment by bufdev** [comment]
See https://github.com/bufbuild/rules_buf for the work so far
